src/imageops/filter_1d.rs

```diff
@@ -0,0 +1,879 @@
+            &mut buffer[start_ky * full_width..(start_ky + 1) * full_width],
+            &scanned_row_kernel,
+        );
+
+        if y >= half_kernel {
+            let mut brows = vec![image; column_kernel_len];
+
+            for (i, brow) in brows.iter_mut().enumerate() {
+                let ky = (i + start_ky + 1) % column_kernel_len;
+                *brow = &buffer[ky * full_width..(ky + 1) * full_width];
+            }
+
+            let dy = y - half_kernel;
+
+            let dst = &mut destination[dy * full_width..(dy + 1) * full_width];
+
+            filter_symmetric_column::<T, I, N>(&brows, dst, image_size, &scanned_column_kernel);
+        }
+
+        start_ky += 1;
+        start_ky %= column_kernel_len;
+    }
+
+    Ok(())
+}
+
+/// Performs 2D separable convolution on the image
+///
+/// Currently implemented only for symmetrical filters
+///
+/// # Arguments
+///
+/// * `image`: Single plane image
+/// * `destination`: Destination image
+/// * `image_size`: Image size see [FilterImageSize]
+/// * `row_kernel`: Row kernel, *size must be odd*!
+/// * `column_kernel`: Column kernel, *size must be odd*!
+fn filter_1d<T, F, I, const N: usize>(
+    image: &[T],
+    destination: &mut [T],
+    image_size: FilterImageSize,
+    row_kernel: &[F],
+    column_kernel: &[F],
+) -> Result<(), ImageError>
+where
+    T: Copy + AsPrimitive<F> + Default + Send + Sync + KernelTransformer<F, I> + AsPrimitive<I>,
+    F: Default + 'static + Copy,
+    I: ToStorage<T>
+        + Mul<I, Output = I>
+        + Add<I, Output = I>
+        + MulAdd<I, Output = I>
+        + Send
+        + Sync
+        + PartialEq
+        + Default
+        + 'static
+        + Copy,
+    i32: AsPrimitive<F> + AsPrimitive<I>,
+    f64: AsPrimitive<T>,
+{
+    if image.len() != image_size.width * image_size.height * N {
+        return Err(ImageError::Limits(LimitError::from_kind(
+            LimitErrorKind::DimensionError,
+        )));
+    }
+    if destination.len() != image_size.width * image_size.height * N {
+        return Err(ImageError::Limits(LimitError::from_kind(
+            LimitErrorKind::DimensionError,
+        )));
+    }
+
+    assert_ne!(row_kernel.len() & 1, 0, "Row kernel length must be odd");
+    assert_ne!(
+        column_kernel.len() & 1,
+        0,
+        "Column kernel length must be odd"
+    );
+
+    let mut scanned_row_kernel = row_kernel
+        .iter()
+        .map(|&x| T::transform(x))
+        .collect::<Vec<I>>();
+    let mut scanned_column_kernel = column_kernel
+        .iter()
+        .map(|&x| T::transform(x))
+        .collect::<Vec<I>>();
+
+    scanned_row_kernel = prepare_symmetric_kernel(&scanned_row_kernel);
+    scanned_column_kernel = prepare_symmetric_kernel(&scanned_column_kernel);
+
+    if scanned_row_kernel.is_empty() || scanned_column_kernel.is_empty() {
+        for (dst, src) in destination.iter_mut().zip(image.iter()) {
+            *dst = *src;
+        }
+        return Ok(());
+    }
```

Comment:
This setup is repeated in the ring_queue version. We should pass it or do the setup after the check, I think I prefer the latter in terms of readability.